C compilation error when passing incompatible type as `T`

### Describe the bug

... this is a reduced example (excluding use of the `toml` module.) of the issue originally opened in #24435 (fixed in #24443 specifically for the `toml` module).

The output from the reduced code, when compiled with `gcc` is this:

```
 v -cc gcc run rpdc_t.v
================== C compilation error (from gcc): ==============
cc: /dev/shm/rpdc_t.01JTWQ5GVRDA4BSR5K5S8GMP2W.tmp.c: In function ‘main__encode_T_Map_string_main__Any’:
cc: /dev/shm/rpdc_t.01JTWQ5GVRDA4BSR5K5S8GMP2W.tmp.c:6989:40: error: expected expression before ‘map’
cc:  6989 |                                 return map[string]main__Any_to_toml(typ);
cc:       |                                        ^~~
=================================================================
```

### Reproduction Steps

This code results in the compile error above:

```v
type Any = string | int

fn (m map[string]Any) to_toml() string {
	mut t := ''
	return t
}

fn main() {
	mut doc := map[string]Any{}
	_ := encode(doc)
}

fn encode[T](typ T) string {
	$for method in T.methods {
		$if method.name == 'to_toml' {
			return typ.$method()
		}
	}
	mp := encode_struct[T](typ)
	return mp.to_toml()
}

fn encode_struct[T](typ T) map[string]Any {
	mut mp := map[string]Any{}
	return mp
}
```

### Expected Behavior

Clean compile and run

### Current Behavior

C error:

```
 v -cc gcc run rpdc_t.v
================== C compilation error (from gcc): ==============
cc: /dev/shm/rpdc_t.01JTWQ5GVRDA4BSR5K5S8GMP2W.tmp.c: In function ‘main__encode_T_Map_string_main__Any’:
cc: /dev/shm/rpdc_t.01JTWQ5GVRDA4BSR5K5S8GMP2W.tmp.c:6989:40: error: expected expression before ‘map’
cc:  6989 |                                 return map[string]main__Any_to_toml(typ);
cc:       |                                        ^~~
=================================================================
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.10 875c165495b0d6b607cb943ced2b3f289903cdef

### Environment details (OS name and version, etc.)

EndeavourOS (rolling, Arch-based)

--- a/vlib/v/gen/c/comptime.v
+++ b/vlib/v/gen/c/comptime.v
@@ -139,7 +139,8 @@ fn (mut g Gen) comptime_call(mut node ast.ComptimeCall) {
 		}
 		return
 	}
-	sym := g.table.sym(g.unwrap_generic(node.left_type))
+	left_type := g.unwrap_generic(node.left_type)
+	sym := g.table.sym(left_type)
 	g.trace_autofree('// \$method call. sym="${sym.name}"')
 	if node.method_name == 'method' {
 		// `app.$method()`
@@ -187,7 +188,7 @@ fn (mut g Gen) comptime_call(mut node ast.ComptimeCall) {
 			}
 		}
 		// TODO: check argument types
-		g.write('${util.no_dots(sym.name)}_${g.comptime.comptime_for_method.name}(')
+		g.write('${g.styp(left_type)}_${g.comptime.comptime_for_method.name}(')
 
 		// try to see if we need to pass a pointer
 		if mut node.left is ast.Ident {
@@ -279,7 +280,7 @@ fn (mut g Gen) comptime_call(mut node ast.ComptimeCall) {
 			}
 			g.write('if (string__eq(${node.method_name}, _SLIT("${method.name}"))) ')
 		}
-		g.write('${util.no_dots(sym.name)}_${method.name}(${amp} ')
+		g.write('${g.styp(left_type)}_${method.name}(${amp} ')
 		g.expr(node.left)
 		g.writeln(');')
 		j++

--- a/vlib/v/gen/c/comptime.v
+++ b/vlib/v/gen/c/comptime.v
@@ -188,7 +188,7 @@ fn (mut g Gen) comptime_call(mut node ast.ComptimeCall) {
 			}
 		}
 		// TODO: check argument types
-		g.write('${g.styp(left_type)}_${g.comptime.comptime_for_method.name}(')
+		g.write('${g.cc_type(left_type, false)}_${g.comptime.comptime_for_method.name}(')
 
 		// try to see if we need to pass a pointer
 		if mut node.left is ast.Ident {
@@ -280,7 +280,7 @@ fn (mut g Gen) comptime_call(mut node ast.ComptimeCall) {
 			}
 			g.write('if (string__eq(${node.method_name}, _SLIT("${method.name}"))) ')
 		}
-		g.write('${g.styp(left_type)}_${method.name}(${amp} ')
+		g.write('${g.cc_type(left_type, false)}_${method.name}(${amp} ')
 		g.expr(node.left)
 		g.writeln(');')
 		j++

--- a/vlib/v/tests/comptime/comptime_call_map_receiver_test.v
+++ b/vlib/v/tests/comptime/comptime_call_map_receiver_test.v
@@ -1,0 +1,26 @@
+type Any = string | int
+
+fn (m map[string]Any) to_toml() string {
+	mut t := ''
+	return t
+}
+
+fn test_main() {
+	mut doc := map[string]Any{}
+	_ := encode(doc)
+}
+
+fn encode[T](typ T) string {
+	$for method in T.methods {
+		$if method.name == 'to_toml' {
+			return typ.$method()
+		}
+	}
+	mp := encode_struct[T](typ)
+	return mp.to_toml()
+}
+
+fn encode_struct[T](typ T) map[string]Any {
+	mut mp := map[string]Any{}
+	return mp
+}

